### PR TITLE
Key filter

### PR DESCRIPTION
**Pull request by Hublerho**
This PR should implement the feature to use key based filtering of auditd events.
Specifying keys in laurel.conf:

[filter]
filter-keys = ["filter-this"]

#24 Currently it only drops the whole message

Looking forward for your feedback

### CODE REVIEW

**Comment by hillu** [comment]
Thank you for your PR!

Could you please rework your commit series so that there are at most a handful of self-contained, linear commits to review?

The `pub filter: bool` you added to the `Coalesce` object should really be part of the `Event`.

**Comment by Hublerho** [comment]
Hi,
I reworked the series and added the `filter: bool` to the event.
Assuming an event should not be filtered upon creation the default is set to false.
Only while processing the value will be set accordingly.

I hope this resolves part of the issue.
Excited to hear from you!

File: src/config.rs [hunk 1]
```diff
@@ -134,26 +137,41 @@ impl std::fmt::Display for Config {
...
     use super::*;
-    use std::path::Path;
+    use std::{path::Path, fs};
 
     #[test]
     fn simple() {
         let c: Config = toml::de::from_str(r#"
 user = "somebody"
-directory = "/path/to/somewhere"
+directory = "/etc/laurel/config.toml"
```
Comment (hillu) [inline src/config.rs]: I think there's a misconception here. directory is meant to be the the "working directory", i.e. the base directory for log files.

File: src/config.rs [hunk 2]
```diff
@@ -134,26 +137,41 @@ impl std::fmt::Display for Config {
...
 statusreport-period = 86400
 [auditlog]
 file = "somefile"
 read-users = ["splunk"]
 "#).unwrap();
         println!("{:#?}", &c);
         assert_eq!(c.user, Some("somebody".to_string()));
-        assert_eq!(c.directory, Some(Path::new("/path/to/somewhere").to_path_buf()));
+        assert_eq!(c.directory, Some(Path::new("/etc/laurel/config.toml").to_path_buf()));
```
Comment (hillu) [inline src/config.rs]: see above

File: src/config.rs [hunk 3]
```diff
@@ -134,26 +137,41 @@ impl std::fmt::Display for Config {
...
         assert_eq!(c.auditlog, Logfile{
             file: Path::new("somefile").to_path_buf(),
             users: Some(vec!["splunk".to_string()]),
             size: None, generations: None,
         });
     }
+// Run if you want to see wether or not your config file is parsed correctly
+    #[test]
+    fn simple_from_file() {
+        let config_file = fs::read_to_string("/etc/laurel/config.toml");
```
Comment (hillu) [inline src/config.rs]: This is the line that broke the CI test.

If you feel it's necessary to parse content from a file, please include it in the `testdata` subdirectory. Most likely, howver, you should be able to get away with parsing from an inline string (see `simple` test).

File: src/main.rs [hunk 4]
```diff
@@ -262,6 +262,10 @@ fn run_app() -> Result<(), Box<dyn Error>> {
 
     coalesce.populate_proc_table()
         .map_err(|e| format!("populate proc table: {}", e))?;
+    print!("{:?}",config.filter.filter_keys);
```
Comment (hillu) [inline src/main.rs]: The `println!` looks like debug output that you forgot to strip.

File: src/coalesce.rs [hunk 5]
```diff
@@ -570,14 +580,16 @@ impl<'a> Coalesce<'a> {
         }
     }
 
+
```
Comment (hillu) [inline src/coalesce.rs:583]: Format-only changes should go into a separate commit.

**Review by hillu — CHANGES_REQUESTED** [review]
The core of your changes look okay now. There are only a few minor issues I'd like to be changed, after that I'd be happy to accept the PR.

Please put formatting-only changes into a separate commit. They distract from the actual functionality you added.

**Comment by hillu** [comment]
btw, would you also like to do the the `filter.filter-labels` feature (as discussed in #24)?

**Comment by Hublerho** [comment]
I reworked/cleaned up the commits, sorry for the mess a lot of the code was due to improper cleanup before commiting (debug code).
Hope it is ok now, otherwise I am happy to correct more :D

**Comment by Hublerho** [comment]
#24 sounds interesting, I can give it a try

**Comment by hillu** [comment]
After rebasing onto `master` and squashing "Test implementation" and "Cleanup", I have pushed your changes.

Thank you very much for your contribution!

How would you like to be credited in the next release notes?

**Comment by Hublerho** [comment]
Great, thank you!

You can just tag me along with the feature if this is ok?

**Pull request closed without merging**